### ISSUE TITLE
"Stuck" keystroke while waiting for autocomplete popup

### ISSUE THREAD

**Issue by tylerb** [Microsoft/vscode-go#1272]
No idea what is causing this or how to start debugging it, but here's what happens:

1. Type a few characters
2. Everything stops responding. Can't type anymore or do anything at all.
3. Last key typed repeats itself about 30 times
4. Autocomplete popup happens

If I completely disable autocomplete:

```
    "editor.quickSuggestions": {
        "other": false,
        "comments": false,
        "strings": false
    }
```

The issue disappears. New issue though: now I don't have autocomplete. :)

**Comment by tylerb** [comment]
Update on this: I recently upgraded my Go version and it seems gocode didn't like that very much.

I deleted everything under $GOPATH/pkg, reinstalled gocode with `go install -a`, then reinstalled all my packages with `go install -a && go test -i`. This seems to have "fixed" the issue.

However, there does remain an issue with vscode: if `gocode` is having trouble or takes a while to respond, VSCode locks up and causes my system to lock up also (though that part might be gocode). Not sure if this is a VSCode issue or a vscode-go issue.

**Comment by ramya-rao-a** [comment]
> Last key typed repeats itself about 30 times

That scares me

> if gocode is having trouble or takes a while to respond, VSCode locks up 

We could kill the process that calls `gocode` if it takes say more than 3 seconds.
And if this happens often, then we could disable completion and notify the user to check whats wrong with gocode.

Thoughts?

**Comment by tylerb** [comment]
I'm thinking that perhaps the issue might be gocode chewing up every single free cycle, slowing my system to a crawl and causing everything to behave oddly. I watched my processors the last time it happened and they all spiked to 100%.

However, that still shouldn't cause VSCode to repeat the last keypress as far as I'm concerned. I get the feeling it registers the key down event, gets stuck waiting on gocode, then registers the keyup event. Perhaps it thinks the key is being held? I don't know.

Regardless, I think killing the gocode process would be a decent way to approach this. It may even be worth adding a "fix it for me" popup or something like that. Basically it just cleans out the `$GOPATH/pkg` directory so everything has to be rebuilt and gocode is happy again.

**Comment by tylerb** [comment]
A bit more here: I just did `"go.gocodeAutoBuild": false` in my settings and everything is fine now. Not quite sure what is happening, but maybe it's just because I have a large project.

Regardless, it's smooth now. I'll just have to manually build once in a while.

**Comment by ramya-rao-a** [comment]
Thanks @tylerb for getting back.

**Comment by tylerb** [comment]
To be fair, the issue still exist. The key does get stuck while gocode is building. It seems that should not happen regardless of what gocode is doing

**Comment by ramya-rao-a** [comment]
Sorry about that.
Yes, that's true. We should have a timeout on the process that is spawned to run gocode

**Comment by tylerb** [comment]
Would it be possible for us not to block keyboard events while waiting for gocode to respond? I feel like that would be a better solution than killing the process.

Then again there could be details there of which I am not aware.

**Comment by ramya-rao-a** [comment]
I'll have to get a repro to understand whats happening behind the scenes
I'll take a look when I get some time

**Comment by tylerb** [comment]
Looks like this is not be the fault of VSCode or the extension. I found https://github.com/tekezo/Karabiner-Elements/issues/545 and it appears to be the cause of my issue. The spike in load caused by `gocode` triggered this issue for me.

**Comment by ramya-rao-a** [comment]
Wow, I would have never guessed that
Thanks for looping back @tylerb
